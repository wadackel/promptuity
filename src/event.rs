//! A module to read events.
//!
//! # Event
//!
//! The `event` module provides functionalities used in prompt handlers.
//! The main offering is the direct re-export of [`crossterm::event`].
//!
//! ## Examples
//!
//! ```ignore
//! use promptuity::event::{KeyCode, KeyModifiers};
//!
//! impl<W: std::io::Write> Prompt<W> for YourPrompt {
//!   // ...
//!   fn handle(&mut self, code: KeyCode, modifiers: KeyModifiers) -> PromptState {
//!     match (code, modifiers) {
//!       (KeyCode::Enter, KeyModifiers::NONE) => PromptState::Submit,
//!       _ => PromptState::Active,
//!     }
//!   }
//!   // ...
//! }
//! ```

/// [`KeyCode`] re-exports from [`crossterm::event`].
pub use crossterm::event::KeyCode;
/// [`KeyModifiers`] re-exports from [`crossterm::event`].
pub use crossterm::event::KeyModifiers;
