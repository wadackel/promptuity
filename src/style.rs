//! A module to apply attributes and colors on your text.
//!
//! # Style
//!
//! The `style` module offers functionality to apply styling to your text.
//! In promptuity, it exposes a simple API that wraps the features of [`crossterm::style`].
//!
//! ## Examples
//!
//! This is an example of styling text.
//!
//! ```
//! use promptuity::style::{Color, Styled};
//!
//! let styled_text = Styled::new("Colored text").fg(Color::Green).bold().to_string();
//!
//! assert_eq!(styled_text, "\u{1b}[38;5;10m\u{1b}[1mColored text\u{1b}[0m");
//! ```
//!
//! This is an example of changing the displayed text based on whether Unicode is supported or not.  
//! Useful for representing icons and other symbols.
//!
//! ```no_run
//! use promptuity::style::Symbol;
//!
//! //              Unicode (main)       ASCII (fallback)
//! //                            \     /
//! let unselected_icon = Symbol("◯", "[ ]");
//! let selected_icon = Symbol("◉", "[x]");
//! ```

/// [`Color`] re-exports from [`crossterm::style::Color`].
pub use crossterm::style::Color;
use crossterm::style::{Attribute, Attributes, ContentStyle, Stylize};

/// A styling utility for strings wrapped in [`crossterm::style::ContentStyle`].
#[derive(Debug)]
pub struct Styled {
    content: String,
    fg: Option<Color>,
    bg: Option<Color>,
    attr: Option<Attributes>,
}

impl Styled {
    pub fn new(content: impl std::fmt::Display) -> Self {
        Self {
            content: content.to_string(),
            fg: None,
            bg: None,
            attr: None,
        }
    }

    pub fn fg(&mut self, color: Color) -> &mut Self {
        self.fg = Some(color);
        self
    }

    pub fn bg(&mut self, color: Color) -> &mut Self {
        self.bg = Some(color);
        self
    }

    pub fn bold(&mut self) -> &mut Self {
        self.attr(Attribute::Bold)
    }

    pub fn italic(&mut self) -> &mut Self {
        self.attr(Attribute::Italic)
    }

    pub fn underline(&mut self) -> &mut Self {
        self.attr(Attribute::Underlined)
    }

    pub fn dim(&mut self) -> &mut Self {
        self.attr(Attribute::Dim)
    }

    pub fn rev(&mut self) -> &mut Self {
        self.attr(Attribute::Reverse)
    }

    pub fn blink(&mut self) -> &mut Self {
        self.attr(Attribute::SlowBlink)
    }

    fn attr(&mut self, attr: Attribute) -> &mut Self {
        match self.attr {
            Some(ref mut attrs) => attrs.set(attr),
            None => self.attr = Some(Attributes::from(attr)),
        }
        self
    }
}

impl std::fmt::Display for Styled {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut style = ContentStyle::new();
        if let Some(fg) = self.fg {
            style = style.with(fg);
        }
        if let Some(bg) = self.bg {
            style = style.on(bg);
        }
        if let Some(attr) = &self.attr {
            style.attributes = *attr;
        }
        write!(f, "{}", style.apply(self.content.as_str()))
    }
}

#[cfg(windows)]
fn is_unicode_supported() -> bool {
    use std::env;

    if env::var("WT_SESSION").is_ok() {
        return true;
    }

    if let Ok(program) = env::var("TERM_PROGRAM") {
        match program.as_str() {
            "vscode" => return true,
            _ => {}
        }
    }

    if let Ok(term) = env::var("TERM") {
        match term.as_str() {
            "xterm-256color" | "alacritty" => return true,
            _ => {}
        }
    }

    false
}

#[cfg(unix)]
fn is_unicode_supported() -> bool {
    true
}

/// A symbol that can be displayed in unicode or ascii.
#[derive(Debug, Copy, Clone)]
pub struct Symbol<'a, 'b>(
    /// The main symbol to display when unicode is supported.
    pub &'a str,
    /// The fallback symbol to display when unicode is not supported.
    pub &'b str,
);

impl<'a, 'b> Symbol<'a, 'b> {
    /// Creates a new [`Symbol`].
    pub fn new(main: &'a str, fallback: &'b str) -> Self {
        Self(main, fallback)
    }
}

impl<'a, 'b> std::fmt::Display for Symbol<'a, 'b> {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        if is_unicode_supported() {
            write!(f, "{}", self.0)
        } else {
            write!(f, "{}", self.1)
        }
    }
}
