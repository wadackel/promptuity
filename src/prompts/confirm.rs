use crate::event::*;
use crate::style::{Color, Styled, Symbol};
use crate::{Prompt, PromptInput, PromptState, RenderPayload};

const S_ACTIVE: Symbol = Symbol("●", ">");
const S_INACTIVE: Symbol = Symbol("○", " ");

/// A trait for formatting the [`Confirm`] prompt.
///
/// All methods have default implementations, allowing you to override only the specific formatting process you need.
///
/// # Examples
///
/// ```no_run
/// use promptuity::prompts::{Confirm, ConfirmFormatter};
///
/// struct CustomFormatter;
///
/// impl ConfirmFormatter for CustomFormatter {
///     fn layout(&self, yes: String, no: String) -> String {
///         format!("{} or {}", yes, no)
///     }
/// }
///
/// let _ = Confirm::new("...").with_formatter(CustomFormatter);
/// ```
pub trait ConfirmFormatter {
    /// Formats the "Yes" option.
    fn yes(&self, active: bool) -> String {
        let icon = if active {
            Styled::new(S_ACTIVE).fg(Color::Green).to_string()
        } else {
            Styled::new(S_INACTIVE).fg(Color::DarkGrey).to_string()
        };
        format!("{} Yes", icon)
    }

    /// Formats the "No" option.
    fn no(&self, active: bool) -> String {
        let icon = if active {
            Styled::new(S_ACTIVE).fg(Color::Green).to_string()
        } else {
            Styled::new(S_INACTIVE).fg(Color::DarkGrey).to_string()
        };
        format!("{} No", icon)
    }

    /// Formats the layout of the active prompt.
    fn layout(&self, yes: String, no: String) -> String {
        format!("{}  /  {}", yes, no)
    }

    /// Formats the submitted value.
    fn submit(&self, value: bool) -> String {
        if value {
            "Yes".into()
        } else {
            "No".into()
        }
    }
}

/// The default formatter for [`Confirm`].
pub struct DefaultConfirmFormatter;

impl DefaultConfirmFormatter {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        Self
    }
}

impl ConfirmFormatter for DefaultConfirmFormatter {}

/// A prompt for inputting a Yes/No choice.
///
/// # Options
///
/// - **Formatter**: Customizes the prompt display. See [`ConfirmFormatter`].
/// - **Hint**: A message to assist with field input. Defaults to `None`.
/// - **Default Value**: The default value of `bool`. Defaults to `false`.
///
/// # Examples
///
/// ```no_run
/// use promptuity::prompts::Confirm;
///
/// let _ = Confirm::new("Are you sure?").with_default(true);
/// ```
pub struct Confirm {
    formatter: Box<dyn ConfirmFormatter>,
    message: String,
    hint: Option<String>,
    value: bool,
}

impl Confirm {
    /// Creates a new [`Confirm`] prompt.
    pub fn new(message: impl std::fmt::Display) -> Self {
        Self {
            formatter: Box::new(DefaultConfirmFormatter),
            message: message.to_string(),
            hint: None,
            value: false,
        }
    }

    /// Sets the formatter for the prompt.
    pub fn with_formatter(&mut self, formatter: impl ConfirmFormatter + 'static) -> &mut Self {
        self.formatter = Box::new(formatter);
        self
    }

    /// Sets the hint message for the prompt.
    pub fn with_hint(&mut self, hint: impl std::fmt::Display) -> &mut Self {
        self.hint = Some(hint.to_string());
        self
    }

    /// Sets the default value for the prompt.
    pub fn with_default(&mut self, value: bool) -> &mut Self {
        self.value = value;
        self
    }
}

impl AsMut<Confirm> for Confirm {
    fn as_mut(&mut self) -> &mut Self {
        self
    }
}

impl Prompt for Confirm {
    type Output = bool;

    fn handle(&mut self, code: KeyCode, modifiers: KeyModifiers) -> PromptState {
        match (code, modifiers) {
            (KeyCode::Enter, _) => PromptState::Submit,
            (KeyCode::Esc, _) | (KeyCode::Char('c'), KeyModifiers::CONTROL) => PromptState::Cancel,
            (KeyCode::Char('y'), KeyModifiers::NONE) | (KeyCode::Char('Y'), KeyModifiers::NONE) => {
                self.value = true;
                PromptState::Submit
            }
            (KeyCode::Char('n'), KeyModifiers::NONE) | (KeyCode::Char('N'), KeyModifiers::NONE) => {
                self.value = false;
                PromptState::Submit
            }
            (KeyCode::Left, _)
            | (KeyCode::Char('h'), _)
            | (KeyCode::Char('p'), KeyModifiers::CONTROL) => {
                self.value = true;
                PromptState::Active
            }
            (KeyCode::Right, _)
            | (KeyCode::Char('l'), _)
            | (KeyCode::Char('n'), KeyModifiers::CONTROL) => {
                self.value = false;
                PromptState::Active
            }
            _ => PromptState::Active,
        }
    }

    fn render(&mut self, state: &PromptState) -> Result<RenderPayload, String> {
        let payload = RenderPayload::new(self.message.clone(), self.hint.clone(), None);

        match state {
            PromptState::Submit => {
                Ok(payload.input(PromptInput::Raw(self.formatter.submit(self.value))))
            }

            PromptState::Cancel => Ok(payload),

            _ => Ok(payload.input(PromptInput::Raw(self.formatter.layout(
                self.formatter.yes(self.value),
                self.formatter.no(!self.value),
            )))),
        }
    }

    fn submit(&mut self) -> Self::Output {
        self.value
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_prompt;

    test_prompt!(
        test_hint,
        Confirm::new("test message").with_hint("hint message"),
        vec![]
    );

    test_prompt!(
        test_default_enter,
        Confirm::new("test message").as_mut(),
        vec![(KeyCode::Enter, KeyModifiers::NONE)]
    );

    test_prompt!(
        test_default_yes,
        Confirm::new("test message").with_default(true),
        vec![(KeyCode::Enter, KeyModifiers::NONE)]
    );

    test_prompt!(
        test_move,
        Confirm::new("test message").as_mut(),
        vec![
            (KeyCode::Left, KeyModifiers::NONE),
            (KeyCode::Right, KeyModifiers::NONE),
            (KeyCode::Char('h'), KeyModifiers::NONE),
            (KeyCode::Char('l'), KeyModifiers::NONE),
            (KeyCode::Char('h'), KeyModifiers::CONTROL),
            (KeyCode::Char('l'), KeyModifiers::CONTROL),
            (KeyCode::Char('p'), KeyModifiers::CONTROL),
            (KeyCode::Char('n'), KeyModifiers::CONTROL),
            (KeyCode::Enter, KeyModifiers::NONE),
        ]
    );

    test_prompt!(
        test_direct_lower_yes,
        Confirm::new("test message").with_default(false),
        vec![(KeyCode::Char('y'), KeyModifiers::NONE)]
    );

    test_prompt!(
        test_direct_upper_yes,
        Confirm::new("test message").with_default(false),
        vec![(KeyCode::Char('Y'), KeyModifiers::NONE)]
    );

    test_prompt!(
        test_direct_lower_no,
        Confirm::new("test message").with_default(true),
        vec![(KeyCode::Char('n'), KeyModifiers::NONE)]
    );

    test_prompt!(
        test_direct_upper_no,
        Confirm::new("test message").with_default(true),
        vec![(KeyCode::Char('N'), KeyModifiers::NONE)]
    );
}
