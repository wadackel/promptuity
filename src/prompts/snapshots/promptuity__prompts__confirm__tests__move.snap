---
source: src/prompts/confirm.rs
---
state: Active
message: test message
hint: none
placeholder: none
input(raw): [38;5;8m○[39m Yes  /  [38;5;10m●[39m No
body(none):
---
state: Active
input(raw): [38;5;10m●[39m Yes  /  [38;5;8m○[39m No
body(none):
---
state: Active
input(raw): [38;5;8m○[39m Yes  /  [38;5;10m●[39m No
body(none):
---
state: Active
input(raw): [38;5;10m●[39m Yes  /  [38;5;8m○[39m No
body(none):
---
state: Active
input(raw): [38;5;8m○[39m Yes  /  [38;5;10m●[39m No
body(none):
---
state: Active
input(raw): [38;5;10m●[39m Yes  /  [38;5;8m○[39m No
body(none):
---
state: Active
input(raw): [38;5;8m○[39m Yes  /  [38;5;10m●[39m No
body(none):
---
state: Active
input(raw): [38;5;10m●[39m Yes  /  [38;5;8m○[39m No
body(none):
---
state: Active
input(raw): [38;5;8m○[39m Yes  /  [38;5;10m●[39m No
body(none):
---
state: Submit
input(raw): No
body(none):
